### PR TITLE
MSSQL and debezium improvements

### PR DESCRIPTION
**Pull request by rodireich**
## What
A few issues in debezium and mssql were discovered as a result of https://github.com/airbytehq/oncall/issues/6192 .
1. Conversion of incoming debezium change events to ChangeEventWithMetadata was not null safe.
2. During an incremental run of mssql, we'd run debezium with configuration that could turn into initial snapshot.
3. We processed incoming debezium events categorized as `op: r` which is can only be part of an initial snapshot.

This PR deals with the discoveries and adds a new troubleshooting page for the root cause of the oncall issue.

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/DebeziumRecordIterator.kt [hunk 1]
```diff
@@ -181,7 +181,7 @@ class DebeziumRecordIterator<T>(
             hasSnapshotFinished = !changeEventWithMetadata.isSnapshotEvent
 
             if (isEventLogged) {
-                val source: JsonNode? = changeEventWithMetadata.eventValueAsJson()["source"]
+                val source: JsonNode? = changeEventWithMetadata.eventValueAsJson?.get("source")
```
Comment (rodireich) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/DebeziumRecordIterator.kt:184]: kotlin doesn't have `?[]` unfortunately

**Comment by rodireich** [comment]
@evantahler I'd like to add a troubleshooting doc for mssql, similarly to what we have in other connectors ([e.g](https://github.com/airbytehq/airbyte/blob/master/docs/integrations/destinations/postgres/postgres-troubleshooting.md)).
Is there anything I need to do other than adding the doc?

**Comment by evantahler** [comment]
@rodireich - you've got to add the new doc to the [sidebar nav](https://github.com/airbytehq/airbyte/blob/master/docusaurus/sidebars.js) as well.  I can help if you get stuck!

**Review by postamar — COMMENTED** [review]
My only concern here is the potential for log spam, LGTM otherwise.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/ChangeEventWithMetadata.kt [hunk 2]
```diff
@@ -6,29 +6,32 @@ package io.airbyte.cdk.integrations.debezium.internals
...
-    fun event(): ChangeEvent<String?, String?> {
-        return event
-    }
+private val LOGGER = KotlinLogging.logger {}
 
-    fun eventKeyAsJson(): JsonNode {
-        return eventKeyAsJson
-    }
+class ChangeEventWithMetadata(private val event: ChangeEvent<String?, String?>) {
+    val eventKeyAsJson: JsonNode?
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/ChangeEventWithMetadata.kt:14]: `NullNode` a source of confusion beause it's not a null and deserializing(null) is actually throwing an exception rather retuning it.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/ChangeEventWithMetadata.kt [hunk 3]
```diff
@@ -6,29 +6,32 @@ package io.airbyte.cdk.integrations.debezium.internals
...
+            event
+                .key()
+                ?.let { Jsons.deserialize(it) }
+                .also { it ?: LOGGER.warn { "Event key is null $event" } }
+    val eventValueAsJson: JsonNode?
+        get() =
+            event
+                .value()
+                ?.let { Jsons.deserialize(it) }
+                .also { it ?: LOGGER.warn { "Event value is null $event" } }
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/ChangeEventWithMetadata.kt:25]: If we get to the situation where value or key are null then something is seriously misoncifugred, as was the case on the [oncall  ticket](https://github.com/airbytehq/oncall/issues/6192).
I'd rather know what the malformed debezium event looks like. I couldn't figure out the root cause until I was able to get to this information.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/DebeziumRecordIterator.kt [hunk 4]
```diff
@@ -340,8 +340,8 @@ class DebeziumRecordIterator<T>(
          * snapshots) t: truncate, m: message
          */
         fun isEventTypeHandled(event: ChangeEventWithMetadata): Boolean {
-            event.eventValueAsJson()["op"]?.asText()?.let {
-                return it in listOf("c", "u", "d", "r", "t")
+            event.eventValueAsJson?.get("op")?.asText()?.let {
+                return it in listOf("c", "u", "d", "t")
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/DebeziumRecordIterator.kt:344]: correct. an 'r' event is always a row read during snapshot - should never be part of a sync in our current scheme of snapshotting ourselves

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/test/kotlin/io/airbyte/cdk/integrations/debezium/internals/DebeziumRecordIteratorTest.kt [hunk 5]
```diff
@@ -80,7 +80,7 @@ class DebeziumRecordIteratorTest {
         "c, true",
         "u, true",
         "d, true",
-        "r, true",
+        "r, false",
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/test/kotlin/io/airbyte/cdk/integrations/debezium/internals/DebeziumRecordIteratorTest.kt:83]: I'm ok with ensuring we don't handle it.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/ChangeEventWithMetadata.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/ChangeEventWithMetadata.kt:14]: I was going between the three options (nullNode, returning an empty json `{}`, or returning null) and I think that the null we return reflects the fact that the event really had no key.

On a sidenote: nobody is actually using eventKeyAsJson as far as I can tell. not on the mssql path at least. 
The crash we had was due to the fact that we used to deserialize everything upfront. So just turning methods into properties improves things here

**Review by postamar — APPROVED** [review]

**Pull request merged**